common/spec/dependabot/git_commit_checker_spec.rb

```diff
@@ -161,9 +161,11 @@
       end
 
       context "when the source code can't be found" do
+        let(:first_instance) { instance_double(DummyPackageManager::MetadataFinder) }
+
         before do
-          allow_any_instance_of(DummyPackageManager::MetadataFinder)
-            .to receive(:look_up_source).and_return(nil)
+          allow(DummyPackageManager::MetadataFinder).to receive(:new).and_return(first_instance)
```

Comment:
Stubbing `.new` on a class is generally considered bad practice for several reasons:

1. **Hiding Object Construction:**
   - When you stub `.new`, you hide the construction of objects. This means any logic in the constructor (`initialize` method) won't be executed, which can lead to missing out on testing important initialization behavior.

2. **Overbroad Stubbing:**
   - Stubbing `.new` affects all instances of the class throughout the test. This is very similar to using `allow_any_instance_of`, which affects all instances of a class, making it difficult to reason about what is being tested.

3. **Test Reliability and Maintenance:**
   - It makes tests less reliable because it’s not clear what parts of the class are actually being tested. Future changes in the class constructor might go unnoticed in tests, leading to brittle tests that are hard to maintain and debug.

4. **Encourages Poor Design:**
   - Relying on stubbing `.new` can indicate poor design where dependencies are not properly injected. Proper dependency injection makes classes easier to test and promotes better design practices.

### Relation to `allow_any_instance_of`

`allow_any_instance_of` suffers from similar issues:

1. **Affecting All Instances:**
   - Just like stubbing `.new`, `allow_any_instance_of` affects all instances of a class, which can lead to unintended consequences and makes tests less predictable and harder to debug.

2. **Global Stubbing Scope:**
   - Both approaches make it unclear which specific instances are being stubbed or mocked, leading to a more global scope of stubbing than necessary. This reduces the test’s precision and clarity.

### Example Solution with Dependency Injection

Instead of stubbing `.new` or using `allow_any_instance_of`, prefer dependency injection to make the code more testable and the tests more precise.

#### Refactored Code Example

Here’s an example of how to refactor the `Dependabot::GitCommitChecker` class and its tests to use dependency injection, avoiding the need to stub `.new` or use `allow_any_instance_of`.

**Class Under Test (`Dependabot::GitCommitChecker`):**

```ruby
class Dependabot::GitCommitChecker
  def initialize(metadata_finder: DummyPackageManager::MetadataFinder.new)
    @metadata_finder = metadata_finder
  end

  def some_method
    source_url = @metadata_finder.look_up_source
    # existing implementation
  end
end
```

**RSpec Test:**

```ruby
require 'spec_helper'

RSpec.describe Dependabot::GitCommitChecker do
  let(:stubbed_metadata_finder) { instance_double(DummyPackageManager::MetadataFinder) }
  let(:git_commit_checker) { described_class.new(metadata_finder: stubbed_metadata_finder) }
  let(:source_url) { "https://bitbucket.org/gocardless/business" }
  let(:upload_pack_fixture) { "business" }
  let(:service_pack_url) do
    "https://bitbucket.org/gocardless/business.git/info/refs?service=git-upload-pack"
  end
  let(:bitbucket_url) do
    "https://api.bitbucket.org/2.0/repositories/gocardless/business/commits/?exclude=v1.5.0&include=df9f605"
  end
  let(:returned) { Dependabot::Source.from_url("fred") }

  before do
    allow(stubbed_metadata_finder).to receive(:look_up_source).and_return(source_url)
    stub_request(:get, service_pack_url)
      .to_return(
        status: 200,
        body: fixture("git", "upload_packs", upload_pack_fixture),
        headers: { "content-type" => "application/x-git-upload-pack-advertisement" }
      )
  end

  context "when the source code can't be found" do
    before do
      allow(stubbed_metadata_finder).to receive(:source_url).and_return(nil)
    end

    it { is_expected.to be(false) }
  end

  context "with source code not hosted on GitHub" do
    before do
      allow(stubbed_metadata_finder).to receive(:look_up_source).and_return(source_url)
      stub_request(:get, bitbucket_url)
        .to_return(
          status: 200,
          body: fixture("git", "upload_packs", upload_pack_fixture),
          headers: { "content-type" => "application/x-git-upload-pack-advertisement" }
        )
    end

    context "when not included in a release" do
      before do
        stub_request(:get, bitbucket_url)
          .to_return(
            status: 200,
            body: fixture("git", "upload_packs", upload_pack_fixture),
            headers: { "content-type" => "application/x-git-upload-pack-advertisement" }
          )
      end

      it { is_expected.to be(false) }
    end
  end
end
```

### Benefits of the Suggested Approach

1. **Clear Dependency Management:**
   - By using dependency injection, you explicitly define what dependencies your class requires. This makes the class easier to understand and test.

2. **Precise Stubbing:**
   - You only stub the specific instance that your class under test interacts with, avoiding overbroad stubbing that affects all instances of the class.

3. **Improved Test Reliability:**
   - Tests become more reliable as they only stub the necessary parts of the dependency, ensuring that the actual behavior of the class under test is properly validated.

4. **Encourages Better Design:**
   - Promotes better design practices by encouraging the use of dependency injection, leading to more modular and testable code.